xds/server_options.go

```diff
@@ -0,0 +1,85 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package xds
+
+import (
+	"net"
+
+	"google.golang.org/grpc"
+)
+
+// Experimental
+//
+// Notice: All APIs in this file are EXPERIMENTAL and may be changed or removed
+// in a later release.
+
+// WithServingModeCallback returns a grpc.ServerOption which allows users to
+// register a callback to get notified about serving mode changes.
+//
+// NOTE: The returned ServerOption must *only* be used in a call to
+// xds.NewGRPCServer() and must not be used in a call to grpc.NewServer().
+func WithServingModeCallback(cb ServingModeCallback) grpc.ServerOption {
+	return &smcOption{cb: cb}
+}
+
+// smcOption is a server option containing a callback to be invoked when the
+// serving mode changes.
+type smcOption struct {
+	grpc.ServerOption
+	cb ServingModeCallback
+}
+
+type serverOptions struct {
+	modeCallback ServingModeCallback
+}
+
+// ServingMode indicates the current mode of operation of the server.
+type ServingMode int
+
+const (
+	// ServingModeStarting indicates that the serving is starting up.
+	ServingModeStarting ServingMode = iota
```

Comment:
How is this different from `NotServing`?

Reply:
When we receive a bad listener config which pushes the server to go to `not-serving` mode, we need to invoke the callback. But I wanted to make sure that if we are already in `not-serving` and we get another bad config, I don't want to invoke the callback. So, if the first config that the server receives is a bad config, then we would need to have some other state in the server to help us determine that the callback needs to be invoked this time.

Given how frequently control planes resend the same config, I thought it was important that we not invoke the callback when the state doesn't change. We could have a `bool` or a `grpcsync.Event` in the `listenerWrapper` which maintains state about whether the first config update has been received. If we do that, we avoid this extra state. But I felt this was simpler as `Starting` was the default value of this type.

Reply:
This seems like an implementation detail, then?  Can we just not expose the initial state to users?  `servingModeStarting ServingMode = iota`.

Also, we probably do want to call the callback if the state changes from `NotServing`+`error1` to `NotServing`+`error2`.

Reply:
> Also, we probably do want to call the callback if the state changes from NotServing+error1 to NotServing+error2.

This should presumably be consistent across language implementations; please check with Java/C folks.

Reply:
Checked with Java and they are doing this, and since this is something that the user can easily do as part of their callback implementation, I dropped it too. Thanks.

I got rid of the `ServingModeStarting` from the public API as well.